src/npyffi/objects.rs

```diff
@@ -44,6 +60,216 @@ pub struct PyArray_Descr {
     pub hash: npy_hash_t,
 }
 
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct _PyArray_DescrNumPy2 {
+    pub ob_base: PyObject,
+    pub typeobj: *mut PyTypeObject,
+    pub kind: c_char,
+    pub type_: c_char,
+    pub byteorder: c_char,
+    pub _former_flags: c_char,
+    pub type_num: c_int,
+    pub flags: npy_uint64,
+    pub elsize: npy_intp,
+    pub alignment: npy_intp,
+    pub metadata: *mut PyObject,
+    pub hash: npy_hash_t,
+    pub reserved_null: [*mut std::ffi::c_void; 2],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct _PyArray_LegacyDescr {
+    pub ob_base: PyObject,
+    pub typeobj: *mut PyTypeObject,
+    pub kind: c_char,
+    pub type_: c_char,
+    pub byteorder: c_char,
+    pub _former_flags: c_char,
+    pub type_num: c_int,
+    pub flags: npy_uint64,
+    pub elsize: npy_intp,
+    pub alignment: npy_intp,
+    pub metadata: *mut PyObject,
+    pub hash: npy_hash_t,
+    pub reserved_null: [*mut std::ffi::c_void; 2],
+    pub subarray: *mut PyArray_ArrayDescr,
+    pub fields: *mut PyObject,
+    pub names: *mut PyObject,
+    pub c_metadata: *mut NpyAuxData,
+}
+
+#[cfg(all(feature = "numpy-1", not(feature = "numpy-2")))]
+pub use PyArray_DescrProto as PyArray_Descr;
+
+#[cfg(all(not(feature = "numpy-1"), feature = "numpy-2"))]
+pub use _PyArray_DescrNumPy2 as PyArray_Descr;
+
+#[cfg(all(feature = "numpy-1", not(feature = "numpy-2")))]
+#[allow(non_snake_case)]
+#[inline(always)]
+pub fn PyDataType_ISLEGACY(_dtype: *const PyArray_Descr) -> bool {
+    true
+}
+
+#[cfg(feature = "numpy-2")]
+#[allow(non_snake_case)]
+#[inline(always)]
+pub fn PyDataType_ISLEGACY(dtype: *const PyArray_Descr) -> bool {
```

Comment:
This functions and all the similar ones below must be `unsafe fn` as they have no way of validating the pointer argument, i.e. the caller has to ensure a valid pointer is passed.